src/app/clusters/network-commissioning/network-commissioning-events.cpp

```diff
@@ -0,0 +1,156 @@
+#include "network-commissioning-events.h"
+#include "network-commissioning.h"
+
+#include <app/util/af.h>
+#include <gen/cluster-id.h>
+#include <gen/command-id.h>
+#include <gen/enums.h>
+
+#include <app/InteractionModelEngine.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/Span.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConnectivityManager.h>
+
+#include <platform/CHIPDeviceEvent.h>
+
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD
+#include <platform/ThreadStackManager.h>
+#endif // CHIP_DEVICE_CONFIG_ENABLE_THREAD
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::clusters;
+using namespace chip::app::clusters::NetworkCommissioning;
+using namespace chip::app::clusters::NetworkCommissioning::Internal;
+
+namespace chip {
+namespace app {
+namespace clusters {
+namespace NetworkCommissioning {
+namespace Internal {
+
+chip::NodeId networkCommissionerNodeId = chip::kAnyNodeId;
+ClusterState sClusterState             = ClusterState::kIdle;
+uint8_t sCurrentEnablingNetworkSeq     = kInvalidNetworkSeq;
+SecureSessionHandle sCommissionerSecureSessionHandle;
+
+void SetEnablingNetworkSeq(uint8_t seq)
+{
+    sCurrentEnablingNetworkSeq = seq;
+}
+
+ClusterState GetClusterState()
+{
+    return sClusterState;
+}
+
+void MoveClusterState(ClusterState newState, CHIP_ERROR err)
+{
+    chip::app::CommandSender * commandSender = nullptr;
+    CHIP_ERROR processError                  = CHIP_NO_ERROR;
+    switch (sClusterState)
+    {
+    case ClusterState::kIdle: {
+        switch (newState)
+        {
+        case ClusterState::kEnablingNetwork:
+            // AddEventHandler will do nothing is it is already added.
+            chip::DeviceLayer::PlatformMgr().AddEventHandler(OnNetworkEnableStatusChangeCallback, 0);
+            break;
+        default:
+            break;
+        }
+    }
+    break;
+    case ClusterState::kEnablingNetwork: {
+        if (newState == ClusterState::kIdle)
```

Comment:
So obviously if we were trying to enable a network and transition to idle, that's a failure and we handle it below.

However, are there other illegal transitions here?  With 3 states, we have 3 factorial = 6 possible transitions.  As the reader, it would be nice if instead of default case statements, all possible transitions were accounted for with a comment on why no-op state transitions are legal and should be a no-op.

I also wonder if this needs to be guarded from incorrect operation with multiple initiators.  Is it really correct to allow one controller to transition the cluster form `kIlde` to `kEnablingNetwork`, and a completely different controller to be able to transition from `kEnablingNetwork` to `kScanningNetwork`?

Reply:
Yes, you are right, per discussions in spec, we only need two general state -- Idle and Working, and EnableNetwork, ScanningNetwork etc. describes the detailed task it is working.
So the transition function can be much simplier.

Reply:
Answered above: `kScanningNetwork` is for future development.

Would it make sense to remove `kScanningNetwork` and default cases for now?  Most compilers will call out missing enum cases without a default case as an error.

Removing these would force the future developer to either fully specify all state transitions, or choose to add a default cases.

Reply:
Sorry.  You were very quick in your response and I had not seen that.

So do I presume then that we want to change this PR to just expose `Idle` and `Working` states?  Also, I think this answers my question about colliding controllers acting on the cluster at the same time.  With only two possible states and two possible transitions, it appears the code fully accounts for this type of corner case.

Reply:
Sorry I'm working on something else and finally have bandwidth to take care of this PR.
Yes, although we do not name the Working state, but in current design (see MoveClusterState), all states except Idle are Working state, and one working state cannot move to another working state.
By using a state machine and send responses in state machine state change,we will have a single place that send asynchronous responses.